### PR TITLE
Optimize Counting Target Votes using Ancestor Cache

### PR DESCRIPTION
**Pull request by terenc3t**
This uses #2269 to calculate target votes for fork choice

Getting target vote was a repetitive process and a lot of computation resources were wasted. With a block ancestor cache we only have to look up a distinct block ancestor once, then all the identical votes could use the fetched block ancestor in the future

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Please use draft PRs for in progress work, if possible

File: beacon-chain/blockchain/fork_choice.go [hunk 1]
```diff
@@ -316,10 +319,43 @@ func (c *ChainService) attestationTargets(ctx context.Context, state *pb.BeaconS
...
+	var ancestor *pb.BeaconBlock
+	var err error
+
 	for validatorIndex, targetBlock := range targets {
-		ancestor, err := BlockAncestor(targetBlock, block.Slot, beaconDB)
-		if err != nil {
-			return 0, err
+		if featureconfig.FeatureConfig().EnableBlockAncestorCache {
+			// if block ancestor cache was enabled, use target block hash + ancestor block height as cache key.
+			targetHash, err := hashutil.HashBeaconBlock(targetBlock)
```
Comment (rauljordan) [inline beacon-chain/blockchain/fork_choice.go]: Can this block be abstracted into its own small function? It makes the vote count func quite big and hard to follow

File: shared/featureconfig/config.go [hunk 2]
```diff
@@ -29,8 +29,10 @@ type FeatureFlagConfig struct {
 	EnableComputeStateRoot       bool // EnableComputeStateRoot implementation on server side.
 	EnableCrosslinks             bool // EnableCrosslinks in epoch processing.
 	EnableCheckBlockStateRoot    bool // EnableCheckBlockStateRoot in block processing.
-	EnableHistoricalStatePruning bool // EnableHistoricalStatePruning when updatifinalized states.
+	EnableHistoricalStatePruning bool // EnableHistoricalStatePruning when updating finalized states.
 	EnableCommitteesCache        bool // EnableCommitteesCache for state transition.
+	EnableBlockAncestorCache     bool //EnableBlockAncestorCache for fork choice optimization.
+	EnableBlock                  bool // EnableCommitteesCache for state transition.
```
Comment (prestonvanloon) [inline shared/featureconfig/config.go]: Enable block? what is this? The comment seems outdated too

Comment (terenc3t) [reply shared/featureconfig/config.go]: oops how did that happen.. thx fo catching!

File: beacon-chain/blockchain/fork_choice.go (hunk 1 above)
Comment (terenc3t) [reply beacon-chain/blockchain/fork_choice.go]: will do!
